localstack/services/awslambda/event_source_listeners/sqs_event_source_listener.py

```diff
@@ -0,0 +1,186 @@
+import time
+from typing import Any, Dict, List
+
+from localstack.services.awslambda import lambda_executors
+from localstack.services.awslambda.event_source_listeners.event_source_listener import (
+    EventSourceListener,
+)
+from localstack.services.awslambda.lambda_api import (
+    LOG,
+    get_event_sources,
+    message_attributes_to_lower,
+    run_lambda,
+)
+from localstack.services.awslambda.lambda_executors import InvocationResult
+from localstack.utils.aws import aws_stack
+from localstack.utils.threads import FuncThread
+
+
+class SQSEventSourceListener(EventSourceListener):
+    # SQS listener thread settings
+    SQS_LISTENER_THREAD: Dict = {}
+    SQS_POLL_INTERVAL_SEC: float = 1
+    # Whether to use polling via SQS API (or, alternatively, reactive mode with SQS updates received directly in-memory)
+    # Advantage of polling is that we can delete messages directly from the queue (via 'ReceiptHandle') after processing
+    USE_POLLING = True
+
+    @staticmethod
+    def get_source_type():
+        return "sqs"
+
+    def start(self):
+        if not self.USE_POLLING:
+            return
+        if self.SQS_LISTENER_THREAD:
+            return
+
+        LOG.debug("Starting SQS message polling thread for Lambda API")
+        self.SQS_LISTENER_THREAD["_thread_"] = thread = FuncThread(self._listener_loop)
+        thread.start()
+
+    def get_matching_event_sources(self) -> List[Dict]:
+        return get_event_sources(source_arn=r".*:sqs:.*")
+
+    def process_event(self, event: Any):
+        if self.USE_POLLING:
+            return
+        # feed message into the first listening lambda (message should only get processed once)
+        queue_url = event["QueueUrl"]
+        try:
+            queue_name = queue_url.rpartition("/")[2]
+            queue_arn = aws_stack.sqs_queue_arn(queue_name)
+            sources = get_event_sources(source_arn=queue_arn)
+            arns = [s.get("FunctionArn") for s in sources]
+            source = (sources or [None])[0]
+            if not source:
+                return False
+
+            LOG.debug(
+                "Found %s source mappings for event from SQS queue %s: %s",
+                len(arns),
+                queue_arn,
+                arns,
+            )
+            # TODO: support message BatchSize here, same as for polling mode below
+            messages = event["Messages"]
+            self._process_messages_for_event_source(source, messages)
+        except Exception:
+            LOG.exception(f"Unable to run Lambda function on SQS messages from queue {queue_url}")
+
```

Comment:
Not sure I understand this part or the "reactive" thing here in general. AFAIK the SQS event source listener should always be used via polling?

Reply:
I don't understand it very well either. I didn't write any of the SQS code, I just copied out the existing stuff into it's own file for the sake of organization. I think that the EventSourceListener class is intended to be able to operate in two ways, a way where it is invoked directly or an async way that works through polling. I'm not sure what the use case is for the former method.
I would like to refactor the EventSourceListener class in general because I don't like that under the current implementation each class has to constantly check the state of _every_ event source mapping to see what has changed. I think there should be some sort of event source mapping manager singleton. But I don't want this PR to grow even larger.

Reply:
Good points @dominikschubert @SwatsonCodes . The reactive / event-based mode was intended in the past to reduce the number of threads as much as possible (for performance reasons). In the case of SQS we sort of need to use polling, though, as there's otherwise no easy way to delete processed messages from SQS via the API. 
But as Sam mentioned, there's certainly room for general improvement/optimization in here..! 👍